asynq/tools.py

```diff
@@ -191,14 +191,11 @@ def new_fun(self, *args, **kwargs):
 
             k = cache_key(args, kwargs)
             try:
-                return ConstFuture(instance_cache[k])
+                return instance_cache[k]
             except KeyError:
-                def callback(task):
-                    instance_cache[k] = task.value()
-
-                task = async_fun(self, *args, **kwargs)
-                task.on_computed.subscribe(callback)
-                return task
+                value = yield async_fun(self, *args, **kwargs)
```

Comment:
For what it's worth, we didn't use this implementation strategy in order to avoid creating an additional asynq task. Our previous experience is that creating an asynq task is relatively expensive and in performance critical paths, avoiding task creation can make a significant difference (Manan will remember). However, we didn't benchmark `@acached_per_instance` specifically and correctness is better than performance.

We could also just catch and ignore the error in the callback, right? The error would still be thrown in the right place inside the scheduler.

Reply:
I think catching the error in the callback would be ok but a little messier. Martin pointed out that we'd be raising the same exception object multiple times, which could cause other weird things to happen.

My preference is to push this and change strategies if we see a regression somewhere.